src/shogun/clustering/KMeansBase.cpp

```diff
@@ -0,0 +1,419 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 1999-2008 Gunnar Raetsch
+ * Written (W) 2007-2009 Soeren Sonnenburg
+ * Copyright (C) 1999-2009 Fraunhofer Institute FIRST and Max-Planck-Society
+ */
+
+#include <shogun/clustering/KMeansBase.h>
+#include <shogun/distance/Distance.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/base/Parallel.h>
+
+#ifdef HAVE_PTHREAD
+#include <pthread.h>
+#endif
+
+#define MUSRECALC
+
+using namespace shogun;
+
+CKMeansBase::CKMeansBase()
+: CDistanceMachine()
+{
+	init();
+}
+
+CKMeansBase::CKMeansBase(int32_t k, CDistance* d, bool use_kmpp)
+: CDistanceMachine()
+{	
+	init();
+	m_k=k;
+	set_distance(d);
+	m_use_kmeanspp=use_kmpp;
+}
+
+CKMeansBase::CKMeansBase(int32_t k_i, CDistance* d_i, SGMatrix<float64_t> centers_i)
+: CDistanceMachine()
+{
+	init();
+	m_k = k_i;
+	set_distance(d_i);
+	set_initial_centers(centers_i);
+}
+
+CKMeansBase::~CKMeansBase()
+{
```

Comment:
I didn't really add any new code and the previous CKMeans didn't have any. I came across some new creation of objects in some methods but they were UNREF in the method itself. 
Still I'll run memcheck on the libhshogun example mayb?